### ISSUE TITLE
Conflicts on the archive module

### ISSUE THREAD

**Issue by dalesit** [antoineco/aco-oracle_java#18]
I've run into a conflict with other modules' dependency on camptocamp's archive module, and this module's dependency on the puppetlabs-archive module. 

```
[root@puppet:manifests]# puppet module install aco-oracle_java
Notice: Preparing to install into /etc/puppetlabs/code/environments/production/modules ...
Notice: Downloading from https://forgeapi.puppetlabs.com ...
Error: Could not install module 'aco-oracle_java' (latest)
  Dependency 'puppet-archive' (v0.4.8) would overwrite /etc/puppetlabs/code/environments/production/modules/archive
    Currently, 'camptocamp-archive' (v0.8.1) is installed to that directory
    Use `puppet module install --ignore-dependencies` to install only this module
[root@puppet:manifests]# puppet module uninstall camptocamp-archive
Notice: Preparing to uninstall 'camptocamp-archive' ...
Error: Could not uninstall module 'camptocamp-archive'
  Other installed modules have dependencies on 'camptocamp-archive' (v0.8.1)
    'bfraser/grafana' (v2.5.0) requires 'camptocamp-archive' (>= 0.3.2 <2.0.0)
    Use `puppet module uninstall --force` to uninstall this module anyway

```

The only workaround I can think of is to use a separate environment for the servers that need Oracle-java installing. As far as I can see, the camptocamp-archive module doesn't allow the cookie to be set that allows you to download the archive directly from oracle. Any thoughts? This seems to be a big problem within the puppet community at the moment.

**Comment by antoineco** [comment]
Hey Joel,

Thanks for reporting that. This is a pretty bad design within Puppet indeed, modules with the same name can't live together in the same environment since the namespace (author) is not taken into account.

Playing with environments could be an option but as far as I remember agents can execute only one environment at a time, so you won't be able to include everything. 

Your best luck is probably to fork my module, rename the references to the archive module, and rename the archive module itself as well. Heavy procedure though...

**Comment by dalesit** [comment]
I've created an environment  with minimal modules, specifically for the server that will be requiring this module:

```
puppet module list --environment=Atlassian
/etc/puppetlabs/code/environments/Atlassian/modules
âââ aco-oracle_java (v2.6.3)
âââ arnoudj-sudo (v1.3.0)
âââ josm5-rootsh (v1.0.0)
âââ puppet-archive (v0.4.8)
âââ puppetlabs-apt (v2.2.2)
âââ puppetlabs-concat (v2.1.0)
âââ puppetlabs-ntp (v4.1.2)
âââ puppetlabs-pe_gem (v0.2.0)
âââ puppetlabs-postgresql (v4.7.1)
âââ puppetlabs-stdlib (v4.11.0)
âââ site (???)
âââ thias-sysctl (v1.0.6)
/etc/puppetlabs/code/modules (no modules installed)
/opt/puppetlabs/puppet/modules (no modules installed)
```

So far the manifest is completely basic:

```
class site::confluence {
        include oracle_java
        class { 'postgresql::server': }

}
```

However it is erroring with 
`Error: Could not retrieve catalog from remote server: Error 400 on SERVER: no parameter named 'provider' at /etc/puppetlabs/code/environments/Atlassian/modules/oracle_java/manifests/download.pp:42 on Archive[/usr/java/jre-8u92-linux-x64.tar.gz] at /etc/puppetlabs/code/environments/Atlassian/modules/oracle_java/manifests/download.pp:42 on node testapp-l02.corp.internal`

Any thoughts? The modules have been newly installed into this environment.

Cheers,

Joel

**Comment by dalesit** [comment]
Update - if I restart puppetserver to clear the cache, the first time it runs it errors with:

```
Error: Could not retrieve catalog from remote server: Error 400 on SERVER: Evaluation Error: 
Error while evaluating a Resource Statement, Could not autoload puppet/type/archive: 
Could not autoload puppet/provider/archive/faraday: no such file to load -- 
/etc/puppetlabs/code/environments/production/modules/archive/lib/puppet_x/bodeco/util at 
/etc/puppetlabs/code/environments/Atlassian/modules/oracle_java/manifests/download.pp:42:5 
on node phh-l-jraapp-l02.corp.internal
```

Which I think points to an error with missing faraday - not needed with puppet-archive >5.0.0

**Comment by antoineco** [comment]
Looks like a version conflict indeed, the error points out to the `production` environment instead of `Atlassian`, this should never happen:

```
/etc/puppetlabs/code/environments/production/modules/archive/lib/puppet_x/bodeco/util
```

I tested my module with this setup, and it worked despite the warning, so v0.5 of puppet-archive shouldn't be the culprit:

```
# puppet module list                          
Warning: Module 'puppet-archive' (v0.5.1) fails to meet some dependencies:
  'aco-oracle_java' (v2.6.3) requires 'puppet-archive' (>= 0.4.4 < 0.5.0)
/etc/puppet/modules
├── aco-oracle_java (v2.6.3)
├── puppet-archive (v0.5.1)  invalid
├── puppetlabs-pe_gem (v0.2.0)
└── puppetlabs-stdlib (v4.12.0)
/usr/share/puppet/modules (no modules installed)
```

**Comment by dalesit** [comment]
I think it is more that puppet-archive <0.5.0 has a dependency on faraday, but not for version 0.5.0 and greater. 

I've tried upgrading puppet-archive to 0.5.1 and am now getting the error:

```
Error: Could not retrieve catalog from remote server: Error 400 on SERVER: Evaluation Error: Error 
while evaluating a Resource Statement, Could not autoload puppet/type/archive: Could not autoload 
puppet/provider/archive/curl: Could not find parent provider ruby of curl at 
/etc/puppetlabs/code/environments/Atlassian/modules/oracle_java/manifests/download.pp:42:5 on 
node phh-l-jraapp-l02.corp.internal
```

which at least is not going out of the environment....

**Comment by antoineco** [comment]
But how could a dependency on a gem change the behaviour of the entire module? Whether a version depends on Faraday or not shouldn't change anything, the behaviour should be consistent.

I can't reproduce in a clean environment unfortunately, but I suspect some caching issues of conflicting modules in the global module path.

**Comment by ngeorgieff** [comment]
I have exactly the same issue. If I clear the cache it complains about faraday and then on the second run:

Error: Could not retrieve catalog from remote server: Error 400 on SERVER: no parameter named 'provider' at /etc/puppetlabs/code/environments/lab/modules/oracle_java/manifests/download.pp:42 on Archive[/usr/java/jre-8u92-linux-x64.tar.gz] at /etc/puppetlabs/code/environments/lab/modules/oracle_java/manifests/download.pp:42 on node is-engineering-t01.domain.com

[root@is-engineering-t01 manifests]# puppet module list
/etc/puppetlabs/code/environments/production/modules
├── aco-oracle_java (v2.6.3)
├── aco-tomcat (v1.3.2)
├── nanliu-staging (v1.0.3)
├── puppet-archive (v0.4.8)
├── puppetlabs-concat (v1.2.5)
├── puppetlabs-firewall (v1.8.1)
├── puppetlabs-pe_gem (v0.2.0)
└── puppetlabs-stdlib (v4.12.0)

**Comment by antoineco** [comment]
This confuses me, the documentation of the aforementioned module clearly states that this parameter is available. If anyone has the chance to investigate in his/her own environment I'd be very grateful.

**Comment by antoineco** [comment]
The conversation about the faraday issue continues in #19

**Comment by antoineco** [comment]
@dalesit this should be fixed in `puppet/archive` v1.0.0. Please also update to `aco/oracle_java` v2.7.0 if possible, to avoid warnings on dependency versions.

**Comment by antoineco** [comment]
Workaround provided in 80cbc71dda5001bc408571bcac564a468e19a5c4
